cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll

```diff
@@ -15,3 +19,32 @@ private class Swap extends DataFlowFunction {
     output.isParameterDeref(0)
   }
 }
+
+/**
+ * A `swap` member function that is used as follows:
+ * ```
+ * obj1.swap(obj2)
+ * ```
+ */
+private class MemberSwap extends TaintFunction, MemberFunction, AliasFunction {
+  MemberSwap() {
+    this.hasName("swap") and
+    this.getNumberOfParameters() = 1 and
+    this.getParameter(0).getType().(ReferenceType).getBaseType().getUnspecifiedType() =
+      getDeclaringType()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isParameterDeref(0)
+    or
+    input.isParameterDeref(0) and
+    output.isQualifierObject()
+  }
+
+  override predicate parameterNeverEscapes(int index) { none() }
+
+  override predicate parameterEscapesOnlyViaReturn(int index) { index = 0 }
```

Comment:
Does this combination really make sense? How can a parameter never escape, but also escape via returns? Should `parameterNeverEscapes` specify that both arguments never escape?

Reply:
`parameterNeverEscapes` has an empty implementation (`AliasFunction` is set up in such a way that there has to be a definition here).

Reply:
Oh! The `Never` part of the `parameterNeverEscapes` predicate had me confused. You're right, all is good!

Reply:
Yeah, I find the `AliasFunction` model a bit confusing and always seem to have to read everything twice.